server/aesdsocket.c

```diff
@@ -0,0 +1,275 @@
+    if (rs == SIG_ERR) 
+    {
+        syslog(LOG_USER | LOG_ERR, "Error in registering SIG_ERR.");
+        graceful_exit();
+        return -1;
+    }
+
+    rs = signal(SIGTERM, signal_handler);
+    if (rs == SIG_ERR) 
+    {
+        syslog(LOG_USER | LOG_ERR, "Error in registering SIGTERM.");
+        graceful_exit();
+        return -1;
+    }
+
+    if (argc == 2) 
+    {
+        if (!strcmp(argv[1], "-d")) 
+        {
+            daemonize = true;
+        } 
+        else 
+        {
+            syslog(LOG_USER | LOG_ERR, "Invalid Arguments.");
+            return (-1);
+        }
+    }
+    
+    int fd = open("/var/tmp/aesdsocketdata",O_RDWR | O_CREAT, 0766);
+    if(fd < 0)
+    {
+        syslog(LOG_USER | LOG_ERR, "Error in opening/ creating the file. Error: %s", strerror(errno));
+        graceful_exit();
+        return -1;
+    }
+    close(fd);
+
+    memset(&hints, 0, sizeof(hints));
+
+    hints.ai_family = AF_INET;
+    hints.ai_socktype = SOCK_STREAM;
+    hints.ai_flags = AI_PASSIVE;
+    int getaddrinfo_result = getaddrinfo(NULL, (PORT_NUM), &hints, &result);
+    if (getaddrinfo_result != 0) 
+    {
+        syslog(LOG_USER | LOG_ERR, "ERROR: %s\n", gai_strerror(getaddrinfo_result));
+        graceful_exit();
+        return -1;
+    }
+
+    sfd = socket(result->ai_family, SOCK_STREAM, 0);
+    if (sfd < 0) 
+    {
+        syslog(LOG_USER | LOG_ERR, "Error in creating the socket. Error: %s", strerror(errno));
+        graceful_exit();
+        return -1;
+    }
+
+    if (setsockopt(sfd, SOL_SOCKET, SO_REUSEADDR, &(int){1}, sizeof(int)) < 0) 
+    {
+        syslog(LOG_USER | LOG_ERR, "Error in setting up socket options. Error: %s", strerror(errno));
+        graceful_exit();
+        return -1;
+    }
+
+    if (bind(sfd, result->ai_addr, result->ai_addrlen) < 0) 
+    {
+        syslog(LOG_USER | LOG_ERR, "Error in binding the socket. Error: %s", strerror(errno));
+        graceful_exit();
+        return -1;
+    }
+
+    if (listen(sfd, BACKLOG)) 
+    {
+        syslog(LOG_USER | LOG_ERR, "Error in making the socket passive. Error: %s", strerror(errno));
+        graceful_exit();
+        return -1;
+    }
+
+    if (daemonize == true) 
+    {
+        int daemon_result = daemon(0,0);
+        if(daemon_result)
+        {
+            syslog(LOG_USER | LOG_ERR, "Error in daemonizing. Error: %s", strerror(errno));
+            graceful_exit();
+            return -1;
+        }
+      }
+
+    while(1) 
+    {
+        cfd = accept(sfd, (struct sockaddr*)&addr, &addr_len);
+        if(cfd < 0)
+        {
+            syslog(LOG_USER | LOG_ERR, "Error in accepting new connection. Error: %s", strerror(errno));
+            graceful_exit();
+            return -1;
+        }
+
+        inet_ntop(AF_INET, &(addr.sin_addr), dst_str, INET_ADDRSTRLEN);
+
+        syslog(LOG_USER | LOG_INFO, "Accepted Connection from %s",dst_str);
+
+        while(1)
+        {
+            bytes_read = read(cfd, buf, 1024);
```

Comment:
@darshitagrawal what will happen if the packet is more than 1024 bytes before you receive your first '\n'? also does this comply with the requirement as stated in  https://www.coursera.org/learn/advanced-embedded-software-development/supplement/N9ZTJ/assignment-5-part-1-instructions which says that the file should be written after newline? This implementation would not work for assignment 8